com.unity.visualeffectgraph/Editor/Data/VFXDataParticle.cs

```diff
@@ -191,8 +192,188 @@ public BucketInfo[] GetBucketLayoutInfo()
+    // Currently we maintain this serialized collection of render state properties due to the following reasons:
+    //   1) Due to the native VFX importer overriding it, we cannot write the full .mat as a sub-asset from C# and store a reference.
+    //   2) Currently there is no concept of a serialized material on the C++ side, it is instantiated on the fly.
+    // Because of these reasons, we instead do the following to inject render state per-context:
+    //   1) Maintain a map of output contexts and materials.
+    //   2) Serialize a truncated version of the material that describe the render state (properties, keywords).
+    //   3) When "compiling" shaders, update the material and sync the render state if necesarry
+    //   4) When building the runtime data, extract the material render state and pack it into a VFXMapping (HACK).
+    // In the future, the full material will be serialized in C++ and this functionality removed.
+    [Serializable]
+    class VFXMaterialCollection : ISerializationCallbackReceiver
+    {
+        private const string k_RenderSettingsMaterialName = "Render Settings";
+
+        private Dictionary<VFXContext, Material> m_MaterialMap = new Dictionary<VFXContext, Material>();
+        private Dictionary<VFXContext, RenderStateProperties> m_RenderStatePropertyMap = new Dictionary<VFXContext, RenderStateProperties>();
+
+        public Material GetOrCreate(VFXContext context)
+        {
+            Material mat;
+
+            if (!m_MaterialMap.TryGetValue(context, out mat))
+            {
+                mat = new Material(VFXResources.defaultResources.shader) { name = k_RenderSettingsMaterialName };
+                m_MaterialMap.Add(context, mat);
+            }
+            else if (mat == null)
+            {
+                mat = new Material(VFXResources.defaultResources.shader) { name = k_RenderSettingsMaterialName };
+                m_MaterialMap[context] = mat;
+            }
+
+            return mat;
+        }
+
+        public void TrySyncRenderStateProperties(VFXContext context)
+        {
+            if (!m_MaterialMap.ContainsKey(context) || !m_RenderStatePropertyMap.ContainsKey(context))
+                return;
+
+            var material = m_MaterialMap[context];
+            var renderStateProperty = m_RenderStatePropertyMap[context];
+
+            // Configure this material properties to the serialized render state cache.
+            renderStateProperty.ConfigureMaterial(material);
+        }
+
+        [Serializable]
+        class RenderStateProperties
+        {
+            [SerializeField]
+            private List<string> m_PropertyNames = new List<string>();
+
+            [SerializeField]
+            private List<float> m_PropertyValues = new List<float>();
+
+            public RenderStateProperties(Material material)
+            {
+                var shader = material.shader;
```

Comment:
This access sometime generates this exception : 
MissingReferenceException: The object of type 'Material' has been destroyed but you are still trying to access it.
Your script should either check if it is null or you should not destroy the object.
UnityEditor.VFX.VFXMaterialCollection+RenderStateProperties..ctor (UnityEngine.Material material) (at G:/Unity/Dev_VFX_Extra/com.unity.visualeffectgraph/Editor/Data/VFXDataParticle.cs:255)
UnityEditor.VFX.VFXMaterialCollection.CacheRenderStates () (at G:/Unity/Dev_VFX_Extra/com.unity.visualeffectgraph/Editor/Data/VFXDataParticle.cs:295)
UnityEditor.VFX.VFXMaterialCollection.OnBeforeSerialize () (at G:/Unity/Dev_VFX_Extra/com.unity.visualeffectgraph/Editor/Data/VFXDataParticle.cs:285)
UnityEditor.EditorApplication:Internal_CallUpdateFunctions()
Same, it should simpler once the material is handled in VFXAsset.